src/util.h

```diff
@@ -112,6 +108,64 @@ static inline bool error(const char* format)
     return false;
 }
 
+/* One of these structures is statically allocated at the site of every call to
+ * LogPrint(category, format, ...), and also at the site of every call to
+ * LogAcceptCategory(category). The first time a site is reached, its structure
+ * is initialized and linked into the global linked list logFastReached (this
+ * "discovery" event is the first time it's possible to do this).
+ */
+struct LogFastEntry {
+    const char* category;
+    const char* line;   // pathname:lineNumber
+    bool enabled;
+    LogFastEntry *next;
+};
+
+/* This function is called the first time (and only the first time) a call
+ * site to LogPrint() or LogAcceptCategory() is visited.
+ */
+void LogFastEntryInit(LogFastEntry *lfe, const char *category, const char* line);
+
+/**
+ * Print the given string (with arguments) if the category is enabled, or if
+ * the line (where the LogPrint occurs) is enabled. This is very efficient
+ * except possibly for the first call in the life of the process.
+ */
+#define LogPrint(category, ...) do {                            \
+    static_assert((category) != nullptr);                       \
+    static LogFastEntry lfe;                                    \
+    static boost::once_flag once = BOOST_ONCE_INIT;             \
+    boost::call_once(                                           \
+        std::bind(&LogFastEntryInit, &lfe,                      \
+            (category),                                         \
+            __FILE__ ":" BOOST_PP_STRINGIZE(__LINE__)),         \
+        once);                                                  \
+    if (lfe.enabled) {                                          \
```

Comment:
I just noticed something about the way I formatted this. Where I came from we always used squiggly braces after an `if` (or `while` etc.) even if it's just one line. Is that okay if I do that here, or should I get rid of the braces? I did this completely out of habit. One thing I really like about our coding standards, btw, is that the open brace is on the same line as the `if`.